### PR TITLE
Adding support for RTL languages

### PR DESCRIPTION
**Pull request by mohataher**
### Identify the Bug

https://github.com/atom/atom/issues/10132 and all their related RTL issues.
https://github.com/atom/atom/issues/17276
https://github.com/atom/atom/issues/15628

### Description of the Change

This is an initial push of the changes required for Atom to support RTL languages. 

Here's the existing behaviour of Atom with RTL languages:
QfBR0

Now here's it working with RTL after the submitted changes:
ezgif-4-dfed6482fbe4

You can clearly tell that it's not fully done yet for merging into the master branch. There are many bugs/features to be added here to make it fully functioning. 

So here's why I'm opening this pull request before code is ready:
1. RTL are either closed because they're too old or locked as too heated. 
2. So there is no (productive) channel to provide a collaborative ground for developers interested in RTL.

I'm hoping this would be a good starting point for all RTL developers to show their skills and contribute to this feature.

To get this done:
1. Please [clone this fork](https://github.com/mohataher/atom) and [run it on your machine](https://flight-manual.atom.io/hacking-atom/sections/contributing-to-official-atom-packages/). 
2. Please fill in the list of current issues that should be solved. Add them here in a comment. A GIF would be great.
3. Please contribute to the code and fix this issue. Please [clone this fork](https://github.com/mohataher/atom) and submit PR to it. 
4. Please write test cases to cover up the following, RTL only text or LTR and RTL text in the same line on LTR/RTL text editor.

### Possible Drawbacks

The code changes enforces RTL on the text editor and won't work on TRL. As started above, this code is not ready for merge just yet and needs further collaboration.

### Verification Process

TBD

### Release Notes

- Text editor can now handle RTL text - still few bugs to fix before merging it.

### CODE REVIEW

File: src/text-editor-component.js [hunk 1]
```diff
@@ -2652,16 +2655,25 @@ module.exports = class TextEditorComponent {
         const textNodeEndColumn =
           textNodeStartColumn + textNode.textContent.length;
 
+        // set clientPixelPosition value --
         if (nextColumnToMeasure < textNodeEndColumn) {
           let clientPixelPosition;
           if (nextColumnToMeasure === textNodeStartColumn) {
-            clientPixelPosition = clientRectForRange(textNode, 0, 1).left;
+            clientPixelPosition = clientRectForRange(textNode, textNode.textContent.length-1, textNode.textContent.lengt).right;
```
Comment (UziTech) [inline src/text-editor-component.js:2662]:
```suggestion
            clientPixelPosition = clientRectForRange(textNode, textNode.textContent.length-1, textNode.textContent.length).right;
```

Tests should be written for this

**Comment by mohataher** [comment]
@lkashef , please have a look and invite other members to contribute to it, if possible.

**Comment by darangi** [comment]
Hi @mohataher , Can You fix the lint errors so the CI can be green?

File: spec/fixtures/sample2.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية
+لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية
+لغة عربية and some english
+لغة عربية لغة عربية لغة عربية لغة عربية
+
+لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية
+لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية
+لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية
+لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية
+لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية لغة عربية
```
Comment (aminya) [inline spec/fixtures/sample2.js:14]: Adding some persian to make the test more compelte:
```suggestion
زبان پارسی  زبان پارسی زبان پارسی زبان پارسی زبان پارسی زبان پارسی زبان پارسی
زبان پارسی زبان پارسی زبان پارسی زبان پارسی زبان پارسی زبان پارسی
 پچژگ پچژگ پچژگ پچژگ پچژگ
```

**Comment by aminya** [comment]
@sadick254 Could you leave a comment about the information you need? A tag does not help the author.

**Comment by aminya** [comment]
@mohataher Could you run `script/lint --fix`?

File: src/decoration-manager.js [hunk 3]
```diff
@@ -204,6 +204,7 @@ module.exports = class DecorationManager {
   }
 
   decorateMarker(marker, decorationParams) {
+    console.log('decorateMarker called') ;
```
Comment (aminya) [inline src/decoration-manager.js:207]: There is no point in having these
```suggestion
```

File: src/decoration-manager.js [hunk 4]
```diff
@@ -238,6 +239,7 @@ module.exports = class DecorationManager {
   }
 
   decorateMarkerLayer(markerLayer, decorationParams) {
+    console.log('decorateMarkerLayer called') ;
```
Comment (aminya) [inline src/decoration-manager.js:242]:
```suggestion
```

File: src/text-editor-component.js [hunk 5]
```diff
@@ -1069,6 +1069,7 @@ module.exports = class TextEditorComponent {
   }
 
   addDecorationToRender(type, decoration, marker, screenRange, reversed) {
+    console.log('addDecorationToRender called') ;
```
Comment (aminya) [inline src/text-editor-component.js:1072]:
```suggestion
```

File: src/text-editor-component.js [hunk 6]
```diff
@@ -1122,6 +1123,7 @@ module.exports = class TextEditorComponent {
   }
 
   addLineDecorationToRender(type, decoration, screenRange, reversed) {
+    console.log('addLineDecorationToRender called') ;
```
Comment (aminya) [inline src/text-editor-component.js:1126]:
```suggestion
```

File: src/text-editor-component.js [hunk 7]
```diff
@@ -1175,6 +1177,7 @@ module.exports = class TextEditorComponent {
   }
 
   addHighlightDecorationToMeasure(decoration, screenRange, key) {
+    console.log('addHighlightDecorationToMeasure called') ;
```
Comment (aminya) [inline src/text-editor-component.js:1180]:
```suggestion
```

File: src/text-editor-component.js [hunk 8]
```diff
@@ -1208,6 +1211,12 @@ module.exports = class TextEditorComponent {
   }
 
   addCursorDecorationToMeasure(decoration, marker, screenRange, reversed) {
+    console.log('addCursorDecorationToMeasure') ; 
+    console.log('screenRange : ' , screenRange) ;
```
Comment (aminya) [inline src/text-editor-component.js:1215]:
```suggestion

```

File: src/text-editor-component.js [hunk 9]
```diff
@@ -1208,6 +1211,12 @@ module.exports = class TextEditorComponent {
   }
 
   addCursorDecorationToMeasure(decoration, marker, screenRange, reversed) {
+    console.log('addCursorDecorationToMeasure') ; 
+    console.log('screenRange : ' , screenRange) ; 
+    var temp = screenRange.start ;
```
Comment (aminya) [inline src/text-editor-component.js:1216]:
```suggestion
    const temp = screenRange.start ; 
```

File: src/text-editor-component.js [hunk 10]
```diff
@@ -1327,6 +1336,7 @@ module.exports = class TextEditorComponent {
   }
 
   addTextDecorationToRender(decoration, screenRange, marker) {
+    console.log('addTextDecorationToRender') ;
```
Comment (aminya) [inline src/text-editor-component.js:1339]:
```suggestion
```

File: src/text-editor-component.js [hunk 11]
```diff
@@ -1479,6 +1491,7 @@ module.exports = class TextEditorComponent {
   }
 
   updateHighlightsToRender() {
+    console.log('updateHighlightsToRender') ;
```
Comment (aminya) [inline src/text-editor-component.js:1494]:
```suggestion
```

File: src/text-editor-component.js [hunk 12]
```diff
@@ -2223,10 +2236,15 @@ module.exports = class TextEditorComponent {
     if (scrolled) this.updateSync();
   }
 
+  debugFun(){
+
+  }
```
Comment (aminya) [inline src/text-editor-component.js:2241]:
```suggestion
```

File: src/text-editor-component.js [hunk 13]
```diff
@@ -2223,10 +2236,15 @@ module.exports = class TextEditorComponent {
...
   screenPositionForMouseEvent(event) {
-    return this.screenPositionForPixelPosition(
+     let state =  this.screenPositionForPixelPosition(
       this.pixelPositionForMouseEvent(event)
     );
+   this.debugFun() ;
```
Comment (aminya) [inline src/text-editor-component.js:2246]:
```suggestion
```

File: src/text-editor-component.js [hunk 14]
```diff
@@ -5223,3 +5337,12 @@ function ceilToPhysicalPixelBoundary(virtualPixelPosition) {
     virtualPixelsPerPhysicalPixel
   );
 }
+
+function checkRTL(s){
+  //if(s==" " || s=="")return true ; 
+    var ltrChars    = 'A-Za-z\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02B8\u0300-\u0590\u0800-\u1FFF'+'\u2C00-\uFB1C\uFDFE-\uFE6F\uFEFD-\uFFFF',
```
Comment (aminya) [inline src/text-editor-component.js:5343]:
```suggestion
    const ltrChars    = 'A-Za-z\u00C0-\u00D6\u00D8-\u00F6\u00F8-\u02B8\u0300-\u0590\u0800-\u1FFF'+'\u2C00-\uFB1C\uFDFE-\uFE6F\uFEFD-\uFFFF',
```

File: src/text-editor.js [hunk 15]
```diff
@@ -289,6 +289,7 @@ module.exports = class TextEditor {
     this.decorateMarkerLayer(this.selectionsMarkerLayer, { type: 'cursor' });
     if (!this.isMini()) this.decorateCursorLine();
 
+    console.log(292) ;
```
Comment (aminya) [inline src/text-editor.js:292]:
```suggestion
```

File: src/text-editor.js [hunk 16]
```diff
@@ -354,6 +355,7 @@ module.exports = class TextEditor {
   }
 
   decorateCursorLine() {
+    console.log('decorateCursorLine') ;
```
Comment (aminya) [inline src/text-editor.js:358]:
```suggestion
```

File: src/text-editor.js [hunk 17]
```diff
@@ -2453,6 +2455,8 @@ module.exports = class TextEditor {
     screenRange = Range.fromObject(screenRange);
     const start = this.bufferPositionForScreenPosition(screenRange.start);
     const end = this.bufferPositionForScreenPosition(screenRange.end);
+    //console.log("start : " , start) ; 
+    //console.log("end : " , end) ;
```
Comment (aminya) [inline src/text-editor.js:2459]:
```suggestion
    console.log("start : " , start) ; 
    console.log("end : " , end) ; 
```

File: src/text-editor.js [hunk 18]
```diff
@@ -1425,7 +1427,7 @@ module.exports = class TextEditor {
   logScreenLines(start = 0, end = this.getLastScreenRow()) {
     for (let row = start; row <= end; row++) {
       const line = this.lineTextForScreenRow(row);
-      console.log(row, this.bufferRowForScreenRow(row), line, line.length);
+      //console.log(row, this.bufferRowForScreenRow(row), line, line.length);
```
Comment (aminya) [inline src/text-editor.js:1430]:
```suggestion
      console.log(row, this.bufferRowForScreenRow(row), line, line.length);
```

File: src/text-editor.js [hunk 19]
```diff
@@ -2642,6 +2646,7 @@ module.exports = class TextEditor {
   //
   // Returns the created {Decoration} object.
   decorateMarker(marker, decorationParams) {
+    console.log('decorateMarker called') ;
```
Comment (aminya) [inline src/text-editor.js:2649]:
```suggestion
```

File: src/text-editor.js [hunk 20]
```diff
@@ -2771,6 +2777,7 @@ module.exports = class TextEditor {
   //
   // Returns a {DisplayMarker}.
   markBufferRange(bufferRange, options) {
+    console.log('markBufferRange called') ;
```
Comment (aminya) [inline src/text-editor.js:2780]:
```suggestion
```

File: src/text-editor.js [hunk 21]
```diff
@@ -2805,6 +2812,7 @@ module.exports = class TextEditor {
   //
   // Returns a {DisplayMarker}.
   markScreenRange(screenRange, options) {
+    console.log('markScreenRange called') ;
```
Comment (aminya) [inline src/text-editor.js:2815]:
```suggestion
```

File: src/text-editor.js [hunk 22]
```diff
@@ -2831,6 +2839,7 @@ module.exports = class TextEditor {
   //
   // Returns a {DisplayMarker}.
   markBufferPosition(bufferPosition, options) {
+    console.log('markBufferPosition called' ) ;
```
Comment (aminya) [inline src/text-editor.js:2842]:
```suggestion
```

File: src/text-editor.js [hunk 23]
```diff
@@ -2862,6 +2871,7 @@ module.exports = class TextEditor {
   //
   // Returns a {DisplayMarker}.
   markScreenPosition(screenPosition, options) {
+    console.log('markScreenPosition called') ;
```
Comment (aminya) [inline src/text-editor.js:2874]:
```suggestion
```

File: src/text-editor.js [hunk 24]
```diff
@@ -2896,6 +2906,7 @@ module.exports = class TextEditor {
   //
   // * `id` {Number} id of the marker
   getMarker(id) {
+    console.log('getMarker : ' , id) ;
```
Comment (aminya) [inline src/text-editor.js:2909]:
```suggestion
```

File: src/text-editor.js [hunk 25]
```diff
@@ -3058,6 +3069,7 @@ module.exports = class TextEditor {
   //
   // Returns a {Cursor}.
   addCursorAtBufferPosition(bufferPosition, options) {
+    console.log('addCursorAtBufferPosition called') ;
```
Comment (aminya) [inline src/text-editor.js:3072]:
```suggestion
```

File: src/text-editor.js [hunk 26]
```diff
@@ -3072,6 +3084,7 @@ module.exports = class TextEditor {
   //
   // Returns a {Cursor}.
   addCursorAtScreenPosition(screenPosition, options) {
+    console.log('addCursorAtScreenPosition called ') ;
```
Comment (aminya) [inline src/text-editor.js:3087]:
```suggestion
```

**Comment by sadick254** [comment]
Hey @mohataher. Thank you for putting a lot of effort into this PR. It would certainly be a good addition to Atom. However, we won't be accepting this new feature. The maintenance cost of this new feature is huge and Atom currently doesn't have the capacity to deal with issues and pull requests that come as a result of this addition.

Once again, thank you for all the hard work on this.

**Comment by mohataher** [comment]
No problem @sadick254. I ran out of resources to keep pushing this feature. We ended up forking and using our own solution internally. Will be happy to give support for anyone interested to push this important feature. Please message me if needed.

**Comment by Masoumeh** [comment]
Dear @mohataher, is it possible to help us using your solution? I tried to follow your instructions above, but then the LTR broke. Are there other changes that I have to get?

I appreciate any help.

**Pull request closed without merging**